### ISSUE TITLE
ImageResponse type error on route after updating to 13.5.1

### ISSUE THREAD

**Issue by Dannymx** [vercel/next.js#55604]
### Link to the code that reproduces this issue

https://github.com/Dannymx/shockinglemon.com/tree/develop

### To Reproduce

1. Clone repo
2. next build
3. See type error on GET route handler

### Current vs. Expected behavior

**Current:**
```
app/api/og/route.tsx
Type error: Route "app/api/og/route.tsx" has an invalid export:
  "Promise<NextResponse<unknown> | ImageResponse>" is not a valid GET return type:
    Expected "Response | Promise<Response>", got "Promise<NextResponse<unknown> | ImageResponse>".
      Expected "Promise<Response>", got "Promise<NextResponse<unknown> | ImageResponse>".
        Expected "Response", got "NextResponse<unknown> | ImageResponse".

 ELIFECYCLE  Command failed with exit code 1.
```
**Expected**: no type errors

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
next info    

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:53 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T6020
    Binaries:
      Node: 18.17.1
      npm: 9.6.7
      Yarn: N/A
      pnpm: 8.7.5
    Relevant Packages:
      next: 13.5.1
      eslint-config-next: 13.5.1
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.2.2
    Next.js Config:
      output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, TypeScript (plugin, built-in types)

### Additional context

This started happening after updating to 13.5.1

<sub>[NEXT-1639](https://linear.app/vercel/issue/NEXT-1639/imageresponse-type-error-on-route-after-updating-to-1351)</sub>

**Comment by jackypan1989** [comment]
Me too.

```
 "Promise<NextResponse<{ url: string | null; }> | NextResponse<{ message: any; }> | undefined>" is not a valid GET return type:
[next]     Expected "Response | Promise<Response>", got "Promise<NextResponse<{ url: string | null; }> | NextResponse<{ message: any; }> | undefined>".
[next]       Expected "Promise<Response>", got "Promise<NextResponse<{ url: string | null; }> | NextResponse<{ message: any; }> | undefined>".
[next]         Expected "Response", got "NextResponse<{ url: string | null; }> | NextResponse<{ message: any; }> | undefined".
[next]           Expected "Response", got "undefined".
```

**Comment by smorimoto** [comment]
You should be able to fix that by replacing it with the newly introduced file `opengraph-image.tsx`.

```
app/api/og/route.tsx -> app/opengraph-image.tsx
```

You may need to add some segment config at the same time.

```diff
+ export const alt = 'Open Graph'
+ export const contentType = "image/png";
```

**Comment by smorimoto** [comment]
If you want to use both `opengraph-image.tsx` and `twitter-image.tsx`, you may want to export a fantastic image component from `og.tsx` and make it shared for those 2 files 🙂

**Comment by hipdev** [comment]
Just in case, replace your Request type with NextRequest like:


```typescript
import type { NextRequest } from 'next/server'

...
export async function GET(request: NextRequest) {
...
}
```
I had the same TS error after updating to 13.5.1 but this fix it.

**Comment by icyJoseph** [comment]
> You should be able to fix that by replacing it with the newly introduced file [`opengraph-image.tsx`](https://nextjs.org/docs/app/api-reference/file-conventions/metadata/opengraph-image#generate-images-using-code-js-ts-tsx).
> 
> ```
> app/api/og/route.tsx -> app/opengraph-image.tsx
> ```
> 
> You may need to add some segment config at the same time.
> 
> ```diff
> + export const alt = "Open Graph";
> + export const contentType = "image/png";
> ```

What if `alt` depends on a dynamic parameter?

**Comment by smorimoto** [comment]
> What if alt depends on a dynamic parameter?

@icyJoseph Isn't this line of documentation the answer for you?

> By default, generated images are [statically optimized](https://nextjs.org/docs/app/building-your-application/rendering/server-components#static-rendering-default) (generated at build time and cached) unless they use [dynamic functions](https://nextjs.org/docs/app/building-your-application/rendering/server-components#server-rendering-strategies#dynamic-functions) or uncached data.

**Comment by icyJoseph** [comment]
@smorimoto nope, because I am generating metadata, where I calculate a special alt tag based of the route.

The ImageResponse doesn't accept alt. So I'd have to use the same static alt text. Unless there's a way to export alt as a function that receives the slug, or if I could pass it to the ImageResponse.

I'm also still investigating :)

**Comment by smorimoto** [comment]
@icyJoseph Oh, I get it. Try this: [generateImageMetadata](https://nextjs.org/docs/app/api-reference/functions/generate-image-metadata)

**Comment by alex289** [comment]
I got it to work just by replacing  the import of the ImageResponse from @vercel/og to next/server

**Comment by rodolphoasb** [comment]
Hey, I don't think this is just related to og images. I'm also seeing this error on a completely different route (and this is only happening after migrating to v 13.5.1).

Here's the error on vercel when I'm trying to deploy:

```
Type error: Route "src/app/api/v1/product/createForOtherUser/route.ts" has an invalid export:
--
16:24:57.436 | "Promise<Response \| undefined>" is not a valid POST return type:
16:24:57.436 | Expected "Response \| Promise<Response>", got "Promise<Response \| undefined>".
16:24:57.436 | Expected "Promise<Response>", got "Promise<Response \| undefined>".
16:24:57.436 | Expected "Response", got "Response \| undefined".
16:24:57.436 | Expected "Response", got "undefined".
16:24:57.436 |  
16:24:57.556 | Error: Command "npm run build" exited with 1
```

And here's my code:

```
import { prisma } from "@/lib/prisma";
import { auth } from "@clerk/nextjs";
import { z } from "zod";

const apiRequestValidator = z.object({
  productName: z.string().min(3).max(255),
  productCategory: z.string(),
  productDescription: z.string().min(2).max(1000),
  productPrice: z.number().min(0),
  creatingAsAdmin: z.boolean(),
  userPhone: z.string(),
  userName: z.string(),
});

export type ApiRequest = z.infer<typeof apiRequestValidator>;

export async function POST(req: Request) {
  const body = await req.json();
  const { userId } = auth();

  if (!userId) {
    return new Response("Unauthorized", { status: 401 });
  }

  const {
    productName,
    productCategory,
    productDescription,
    productPrice,
    creatingAsAdmin,
    userName,
    userPhone,
  } = apiRequestValidator.parse(body);

  // Check if user is admin
  const user = await prisma.user.findUnique({
    where: {
      id: userId,
    },
  });

  if (user?.userRole !== "ADMIN") {
    return new Response("Unauthorized", { status: 401 });
  }

  if (creatingAsAdmin) {
    try {
      const result = await prisma.product.create({
       ...
    })

      return new Response(
        JSON.stringify({
          productId: result.id,
        }),
        { status: 200 }
      );
    } catch (error) {
      return new Response("Invalid request", { status: 400 });
    }
  }
}

```

**Comment by Dannymx** [comment]
> I got it to work just by replacing the import of the ImageResponse from @vercel/og to next/server

This fixes it, thanks. I just noticed the docs were updated too.

**Comment by huozhi** [comment]
Hi, can you try to use `import { ImageResponse } from 'next/server'`? We're working on the fix about the type checking now

**Comment by mocca-dev** [comment]
> Just in case, replace your Request type with NextRequest like:
> 
> ```ts
> import type { NextRequest } from 'next/server'
> 
> ...
> export async function GET(request: NextRequest) {
> ...
> }
> ```
> 
> I had the same TS error after updating to 13.5.1 but this fix it.

Hi, I'm having the same problem but with the NextResponse...here are a capture of the error message when I run the build command.

<img width="819" alt="Screenshot 2023-09-20 at 23 11 12" src="https://github.com/vercel/next.js/assets/16454273/9aaff4be-9633-4252-801a-05c3b378004e">

Could you help me with this? thanks

**Comment by anthonyverducci** [comment]
I don't think this is just related to images, either. My API routes were working just fine until I upgraded to `13.5.2` this morning. I'm getting the following error when I deploy to Vercel:

```
Type error: Route "src/app/api/providers/create-provider-bios/route.ts" has an invalid export:
--
09:01:44.392 | "Promise<NextResponse<string> \| undefined>" is not a valid PUT return type:
09:01:44.392 | Expected "Response \| Promise<Response>", got "Promise<NextResponse<string> \| undefined>".
09:01:44.392 | Expected "Promise<Response>", got "Promise<NextResponse<string> \| undefined>".
09:01:44.393 | Expected "Response", got "NextResponse<string> \| undefined".
09:01:44.393 | Expected "Response", got "undefined".
```

In my route, I'm using both `NextRequest` and `NextResponse` imports from `next/server`.

**Comment by balazsorban44** [comment]
Let's open a new issue for these! `ImageResponse` should be correct now

**Comment by balazsorban44** [comment]
That looks slightly different, specifically for `redirect()`.

**Comment by Revaycolizer** [comment]
> I don't think this is just related to images, either. My API routes were working just fine until I upgraded to `13.5.2` this morning. I'm getting the following error when I deploy to Vercel:
> 
> ```
> Type error: Route "src/app/api/search/route.ts" has an invalid export:
> --
> 09:01:44.392 | "Promise<NextResponse<string> \| undefined>" is not a valid PUT return type:
> 09:01:44.392 | Expected "Response \| Promise<Response>", got "Promise<NextResponse<string> \| undefined>".
> 09:01:44.392 | Expected "Promise<Response>", got "Promise<NextResponse<string> \| undefined>".
> 09:01:44.393 | Expected "Response", got "NextResponse<string> \| undefined".
> 09:01:44.393 | Expected "Response", got "undefined".
> ```
> 
> In my route, I'm using both `NextRequest` and `NextResponse` imports from `next/server`.

**Comment by tjhorwood** [comment]
@Revaycolizer @anthonyverducci Not sure if this will help, but I was getting the same exact issue on some of my api routes. It came down to the more complex if/then statements in the catch. For instance changing `catch (error) {
    const errorMessage = 'An error occured!';
    if (error instanceof Error) {
      const isDuplicate = error.message.includes(
        'Unique constraint failed on the fields: (`username`)',
      );
      if (isDuplicate) {
        return NextResponse.json(
          { message: 'Username is already taken, please try another one' },
          { status: 501 },
        );
      } else {
        return NextResponse.json({ message: errorMessage }, { status: 503 });
      }
    }
  }`

**Comment by Revaycolizer** [comment]
But I'm not having any if conditions

**Comment by huozhi** [comment]
We had a fix in 13.5.3, can you try to upgrade to the latest version? Thanks!

**Comment by stevenspads** [comment]
Just updated to 13.5.3 but I'm still getting the following error with an API route:

```
Type error: Type '{ __tag__: "GET"; __return_type__: Promise<string | NextResponse<{ url: string; }> | NextResponse<{ error: string; }>>; }' does not satisfy the constraint '{ __tag__: "GET"; __return_type__: void | Response | Promise<void | Response>; }'.
  Types of property '__return_type__' are incompatible.
    Type 'Promise<string | NextResponse<{ url: string; }> | NextResponse<{ error: string; }>>' is not assignable to type 'void | Response | Promise<void | Response>'.
      Type 'Promise<string | NextResponse<{ url: string; }> | NextResponse<{ error: string; }>>' is not assignable to type 'Promise<void | Response>'.
        Type 'string | NextResponse<{ url: string; }> | NextResponse<{ error: string; }>' is not assignable to type 'void 
| Response'.
          Type 'string' is not assignable to type 'void | Response'.

  57 |         __return_type__: Response | void | never | Promise<Response | void | never>
  58 |       },
> 59 |       {
     |       ^
  60 |         __tag__: 'GET',
  61 |         __return_type__: ReturnType<MaybeField<TEntry, 'GET'>>
  62 |       },
  ```

**Comment by cvogl** [comment]
Unfortunately the patch was not a fix, at least not here. 
tl;dr  Code unchanged since July, breaking of build persists with Next 13.5.3 (but works with 13.4.19)


I have just upgraded to Next 13.5.3 and linter breaks build with the same error message as with 13.5.1:

Type error: Route "src/app/api/frontend/checksession/route.ts" has an invalid export:
  "Promise<false | NextResponse<{ status: string; }>>" is not a valid GET return type:
    Expected "void | Response | Promise<void | Response>", got "Promise<false | NextResponse<{ status: string; }>>".
      Expected "Promise<void | Response>", got "Promise<false | NextResponse<{ status: string; }>>".
        Expected "void | Response", got "false | NextResponse<{ status: string; }>".
          Expected "void | Response", got "boolean".


Double-checked the build after downgrade to 13.4.19: still finishing successfully.
(Which is no surprise: the code in that module was not touched after August 2.)

For us, it's also unrelated to images.

Thank you for investigating!

**Comment by huozhi** [comment]
@cvogl please check the response [here](https://github.com/vercel/next.js/issues/55623#issuecomment-1735889535)
